core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -1478,4 +1478,12 @@ default boolean supportsReportingWrittenBytes(ConnectorSession session, Connecto
     {
         return false;
     }
+
+    /**
+     * Returns true if a connector supports multiple writers to write the same partition concurrently for a partitioned table. Otherwise, it returns false.
+     */
+    default boolean supportsMultipleWritersPerPartition(ConnectorSession session, SchemaTableName schemaTableName)
```

Comment:
"partitioned table" and "partition" are not first-class concepts in Trino SPI, AFAICT, but @martint will know better

The only place where SPI says about "partitioning" is in `ConnectorTableLayout`.
In fact, why not add this new boolean to that class?

cc @losipiuk

Reply:
That's correct. There's no such a concept in the engine at the language or metadata level. Any notion of partitioning is a physical concept for the purpose of query planning.

(I commented on this earlier, not sure why it got lost: https://github.com/trinodb/trino/pull/14718#discussion_r1002757591)

Reply:
The concept of partitioning in the Trino engine is as a way to describe the layout of the data coming out of the connector when reading or writing table -- they are not an entity, as in Hive and Iceberg. You can't create a partition, or describe a partition's metadata. There may not even be a single partitioning scheme (getInsertLayout, getNewTableLayout, getLayoutForTableExecute can all return different partitioning schemes if they want to). 

Hence why a `supportsMultipleWritersPerPartition` at the ConnectorMetadata level makes little sense. What is a "partition" in this context?

Reply:
@martint an alternative would be to add a method like:
```
io.trino.spi.connector.ConnectorPartitioningHandle#isSingleWriterPerPartition
```
However, we concluded that since `ConnectorPartitioningHandle` is not only used for inserts, then separate method like `supportsMultipleWritersPerPartition` is better

Reply:
My question still stands: what is "a partition" in this context, given that the only thing the engine knows about is that data is partitioned based on a certain set of columns but has not concept of a partition being an entity.

Reply:
@sopel39 I would add this to `ConnectorTableLayout`.

This would help us clarify and make visible the now-secret mechanics of that class

- `ConnectorTableLayout#partitionColumns` is optional partitioning (connector must support multiple writers per "partition")
- `ConnectorTableLayout#partitioning` is (today) mandatory partitioning and connector may assume (today) single writer per partition

If we add a new boolean field to that class 

- it will be easier to populate from existing constructors
- it will make the implicit meaning of the fields explicit
- we still will  be able to block `! allowMultipleWriters && partitioning.isEmpty()` case which we don't plan to implement right now (but may choose to implement in the future)

cc @losipiuk @arhimondr

Reply:
We extracted that into https://github.com/trinodb/trino/pull/14956